File: python/dynamic_array/dynamic_array.py
# Pythonic arrays are dynamic by default, this is just an example of resizing.

class DynamicArray:
    def __init__(self):
        self._array = [0] * 1 #Start small, capacity of 2.
        self._length = 0
        self._capacity = 1
    
    def resize(self):
        # Create new array of double capacity
        self._capacity *= 2
        new_array = [0] * self._capacity

        #Copy elements to new_array
        for i in range(self._length):
            new_array[i] = self._array[i]
        self._array = new_array
    
    # Insert n in the last position of the array
    def pushback(self, n):
        if self._length == self._capacity:
            self.resize()
        # Insert at next empty position
        self._array[self._length] = n
        self._length += 1

    # Remove the last element in the array
    def popback(self):
        if self._length > 0:
            self._length -= 1

    # Get value at i-th index
    def get(self, i):
        if i < self._length:
            return self._array[i]
        raise IndexError("Index out of bounds")

    # Insert at the i-th index
    def insert(self, i, n):
        if i < 0 or i > self._length:
            raise IndexError("Index out of bounds")
        if self._length == self._capacity:
            self.resize()
        # Shift right from i to end
        for j in range(self._length, i, -1):
            self._array[j] = self._array[j - 1]
        self._array[i] = n
        self._length += 1
        
    def print(self):
        for i in range(self._length):
            print(self._array[i])
        print()

File: dynamic_array/ts/dynamic_array.ts
export class DynamicArray {
    private _array: number[];
    private _length: number = 0;
    private _capacity: number = 1;

    constructor() {
        this._array = new Array(1).fill(0); // Start small explicit zeros
    }

    private resize(): void {
        this._capacity *= 2;
        const new_array: number[] = new Array(this._capacity).fill(0);
        for (let i = 0; i < this._length; i++) {
            new_array[i] = this._array[i];
        }
        this._array = new_array;
    }

    pushback(n: number): void {
        if (this._length === this._capacity) {
            this.resize();
        }
        this._array[this._length] = n;
        this._length++;
    }

    popback(): void {
        if (this._length > 0) {
            this._length--;
        }
    }

    get(i: number): number {
        if (i < 0 || i >= this._length) {
            throw new RangeError("Index out of bounds");
        }
        return this._array[i];
    }

    insert(i: number, n: number): void {
        if (i < 0 || i > this._length) {
            throw new RangeError("Index out of bounds");
        }
        if (this._length === this._capacity) {
            this.resize();
        }
        for (let j = this._length; j > i; j--) {
            this._array[j] = this._array[j - 1];
        }
        this._array[i] = n;
        this._length++;
    }

    print(): void {
        const output = this._array.slice(0, this._length).join(" ");
        console.log(output || "");
    }

    getLength(): number {
        return this._length;
    }

    getCapacity(): number {
        return this._capacity;
    }
}
